cli/pkg/git/git_utils.go

```diff
@@ -0,0 +1,55 @@
+package git
+
+import (
+	"encoding/json"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+type gitHubApi []struct {
+	Ref    string `json:"ref"`
+	NodeID string `json:"node_id"`
+	URL    string `json:"url"`
+	Object struct {
+		Sha  string `json:"sha"`
+		Type string `json:"type"`
+		URL  string `json:"url"`
+	} `json:"object"`
+}
+
+func CloneGitHubUrl(filePath string, url string, tag string) error {
```

Comment:
would it be possible to directly use the Github API to download the repository content (see https://docs.github.com/en/rest/reference/repos#download-a-repository-archive-tar)? Adding an additional dependency to `github.com/go-git/go-git/v5` seems to be a bit of an overkill if only the `PlainClone` function from this library will be used

Reply:
Using a tar/zip download from GitHub API, would mean that we will have to unzip/untar the downloaded files ourselves. This would involve OS I/O Operations, in the long run, the unzipping/untaring operational logic might fail maybe due to OS constraints or due to how golang handles the filesystem (something which they did in `go1.16` release)

However I also felt that one additional library would be an overkill for the plain clone, but during the research phase, I had found the following pros and cons and had discussed the same.


> ## Research Phase
> ### How to initialise the template?
> There were two options to be explored:
> 
> * Clone the template
>   
>   * Would need to have `git` installed in the developer machine
>   * No need to unzip or download, which may cause issues.
>   * We can either use a system interop call or make use of _git implementation written in go_, [go-git](https://github.com/go-git/go-git)
> * Download and unzip the contents
>   
>   * Might be ineffective on some machines
>   * maintainance issues
>   * does not need `git` as a pre-requiste
> 
> **Conclusion:** Cloning the template and inclusion of the `go-git` as a dependency would be the best way to proceed as we can mock and write _unit-testing_ and would also not require the developer to have `git` installed as default.


**An alternative solution** could be to assume by default that the user has `git` installed. Then we could simply run an `exec` to clone the required repository.

Reply:
thank you for the explanation - those are valid points. In my opinion I'd say it's OK to do this using the go-git library, but I would also like to ask @christian-kreuzberger-dtx and @warber for their opinions on this.

Reply:
@DarthBenro008 I don't quite understand your arguments about why the manual extraction approach involves OS I/O while the other does not. Anyway, I think there is a bigger issue with cloning the repo as you will also receive the entire `.git` folder which is not what we want in this context. 
IMO you can quite easily use the facilities of the `archive/zip` (or `compress/gzip` if you'd rather want to process the tarball ) package to extract the downloaded content. Note, that this can even be done _in-memory_ by creating a `Reader` directly from the response body without temporarily storing the zip/tarball`

``` 
[...]
 zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
[...]
``` 
Given that the template repository is quite small this should be enough.
From then on you can just loop over each `File` entry of the `zipReader` and create the corresponding files and directories on disk. (you'll find plenty of examples for the logic of extracting a zip or tarball correctly online). Note that you might also think about writing the files in a temporary directory in the first place and when everything worked move it over to the final destination.

hope this helps...  :slightly_smiling_face:

Reply:
@warber thank you for the detailed insights, the .git folder is actually removed after the PlainClone as done [here](https://github.com/DarthBenro008/keptn/blob/feat/cli-service-generation/cli/cmd/generate_keptn_service.go#L71)

The `archive/zip` package then should be helpful as we already making use of GitHub API. The only concern was performance. As the repo is small, it should be insignificant. I will try out this approach! 😃